_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,644 @@
+
+Keycloak can be started in development mode with the `start-dev` command or production mode with the `start` command. Production mode follows a _secure by default_ principle and expects _hostname_ and _HTTPS/TLS_ configuration to be set, otherwise, Keycloak won't start and will display an error. Also, in production mode HTTP is disabled by default.
+
+Keycloak documentation provides some key guidelines for production deployment that apply to all environments.
+
+- **HTTPS/TLS**: The exchange of credentials and other sensitive data with Keycloak requires all communication to and from Keycloak to be secured, so HTTP over TLS must be enabled.
+- **Keycloak hostname**: Keycloak usually runs in a private network but certain public-facing endpoints must be exposed to applications. The base URLs determine how tokens are issued and validated, how action links are created, and how the OpenID Connect Discovery Document `realms/{realm-name}/.well-known/openid-configuration` is created.
+- **Reverse proxy**: A reverse proxy/load balancer component is recommended for a production environment, unifying access to the network. Keycloak supports multiple proxy modes. The `edge` mode allows HTTP communication between the proxy and Keycloak, and the proxy keeps a secure connection with clients.
+- **Production grade database**: The database plays a crucial role in the performance and Keycloak supports several production-grade databases, including PostgreSQL.
+- **High Availability**: Choose multi-mode clustered deployment. In production mode, distributed caching of realm and session data is enabled and all nodes in the network are discovered.
+
+## Deploy Spring Boot microservices and Keycloak to Google Kubernetes Engine
+
+In this walkthrough, you will build a microservices architecture example from JHipster, consisting of a `gateway` application based on Spring Gateway and two Spring Boot reactive microservices `blog` and `store`. The `gateway` will act as the entrance to your microservices, providing HTTP routing and load balancing, quality of service, security and API documentation for all microservices. With the help of the JHipster Kubernetes sub-generator, you can deploy the application and its required services (Consul for service discovery, and Keycloak for OpenID Connect) to Google Kubernetes Engine (GKE).
+
+### Build the Spring Boot microservices architecture
+
+Install JHipster, you can do the classical local installation with npm.
+
+```bash
+npm install -g generator-jhipster@7.9.3
+```
+If you'd rather use Yarn or Docker, follow the instructions at [jhipster.tech](https://www.jhipster.tech/installation/#local-installation-with-npm-recommended-for-normal-users).
+
+Generate the microservices architecture from a JDL (JHipster-specific domain language) descriptor file. Create a folder for the project and add the file `reactive-ms.jdl` with the following content:
+
+```text
+application {
+  config {
+    baseName gateway
+    reactive true
```

Comment:
Reactive removed. Tested and it worked.

Reply:
For an application of type `gateway`, it doesn't matter if you specify reactive or not. It will always be reactive because Spring Cloud Gateway only supports WebFlux. This is changing, but JHipster will need to implement a Spring MVC version before it matters. https://twitter.com/spencerbgibb/status/1677547823938641920